### PR TITLE
Openmp profiling

### PR DESCRIPTION
**Pull request by pp1336**
An addition to PR#89, allows profiling under OpenMP settings.

### CODE REVIEW

File: devito/propagator.py [hunk 1]
```diff
@@ -308,7 +308,11 @@ def generate_loops(self, loop_body):
         omp_master = [cgen.Pragma("omp master")] if self.compiler.openmp else []
         omp_single = [cgen.Pragma("omp single")] if self.compiler.openmp else []
         omp_parallel = [cgen.Pragma("omp parallel")] if self.compiler.openmp else []
-        omp_for = [cgen.Pragma("omp for schedule(static)")] if self.compiler.openmp else []
+        reduction_list = ["kernel", "loop_body"] if self.compiler.openmp else []
```
Comment (vincepandolfo) [inline devito/propagator.py]: This should be managed by the Profiler instead of being hard coded.

File: devito/propagator.py [hunk 2]
```diff
@@ -308,7 +308,11 @@ def generate_loops(self, loop_body):
...
+        omp_for = [cgen.Pragma("omp for schedule(static)"\
+                   + (self.profiler.get_loop_reduction("+", reduction_list)\
+                   if self.profile else ""))]\
+                  if self.compiler.openmp else []
```
Comment (vincepandolfo) [inline devito/propagator.py]: This should be managed by the Profiler as well, so that we can do a different reduction on timing since that doesn't work with OpenMP and `+` is not the right reduction here I believe.

File: devito/propagator.py (hunk 1 above)
Comment (pp1336) [inline devito/propagator.py]: Well as we discussed just now, the other parts of the profiling code are also hard coding the these stuff, so I think this is a more general issue that deserves another PR.

File: devito/propagator.py (hunk 2 above)
Comment (pp1336) [inline devito/propagator.py]: the + is only for flops, I did know that timing was not ok with openmp guess a little miscommunication here, I will add the change to handle loop timing

File: devito/profiler.py [hunk 3]
```diff
@@ -1,30 +1,86 @@
...
+        OpenMP settings.
+        :param op: The reduction operator.
+        :param varaiables: A list of string, each is the name of a variable
+                           to be reduced.
+        :returns: String representing the reduction pragma
+        """
+        if len(variables) <= 0:
+            return ""
+        return " reduction(" + op + ":" + self.loop_temp_prefix + variables[0]\
+               + "".join(["," + self.loop_temp_prefix + v for v in variables[1:]]) + ")"
```
Comment (mlange05) [inline devito/profiler.py]: Using `+` to concatenate strings is considered bad style by many. String formatting (`%s`) is much cleaner and more readable.

File: devito/profiler.py [hunk 4]
```diff
@@ -1,30 +1,86 @@
...
+        :param value: String represeting the initial value of temporary
+                      varaibles.
+        :param variables: A list of string, each is the name of a variable to
+                          be declaried
+        :returns: String representing the declariation statement
+        """
+        if len(variables) <= 0:
+            return ""
+        return "long long " + self.loop_temp_prefix + variables[0] + " = " + value +\
+               "".join([", " + self.loop_temp_prefix + v + " = " + value for v in variables[1:]])
```
Comment (mlange05) [inline devito/profiler.py]: String formatting

File: devito/profiler.py [hunk 5]
```diff
@@ -1,30 +1,86 @@
...
+
+    def get_loop_flop_update(self, variables):
+        """Function to generate a list of statement representing the amount
+        of flops update per iteration of the loop.
+        :param variables: A list of string, each is the suffix of the
+                          temperary variables used for update
+        :returns: A list of string representing statements used as updating
+                  statements
+        """
+        return [self.f_name + "->" + v + " += " + self.loop_temp_prefix + v for v in variables]
```
Comment (mlange05) [inline devito/profiler.py]: String formatting

**Comment by mlange05** [comment]
Ok, the code looks great, except for a few minor string formatting comments. One thing I found when running this though is that `gflops` seem to double with two threads, which I don't think is correct. OI and timings look sensible though.

File: devito/profiler.py [hunk 6]
```diff
@@ -1,30 +1,86 @@
...
-    def __init__(self):
-        self.name = "timings"
-        self.fields = []
-        self.timings = None
+    def __init__(self, openmp=False):
+        self.openmp = openmp
+        self.t_name = "timings"
+        self.o_name = "oi"
+        self.f_name = "flops"
+        self.loop_temp_prefix = "temp_"
```
Comment (FabioLuporini) [inline devito/profiler.py]: Feels like ``t_name, o_name, ..., loop_temp_prefix`` should actually be class (read-only) attributes ?
Also, the names you chose are not very explanatory

File: devito/profiler.py [hunk 7]
```diff
@@ -1,30 +1,86 @@
...
+        """
+        if len(variables) <= 0:
+            return ""
+        return "long long " + self.loop_temp_prefix + variables[0] + " = " + value +\
+               "".join([", " + self.loop_temp_prefix + v + " = " + value for v in variables[1:]])
+
+    def get_loop_flop_update(self, variables):
+        """Function to generate a list of statement representing the amount
+        of flops update per iteration of the loop.
+        :param variables: A list of string, each is the suffix of the
```
Comment (FabioLuporini) [inline devito/profiler.py]: Leave a blank line here before ``:param variables: ...``

File: devito/profiler.py [hunk 8]
```diff
@@ -1,30 +1,86 @@
...
+        """
+        if len(variables) <= 0:
+            return ""
+        return " reduction(" + op + ":" + self.loop_temp_prefix + variables[0]\
+               + "".join(["," + self.loop_temp_prefix + v for v in variables[1:]]) + ")"
+
+    def get_loop_temp_var_decl(self, value, variables):
+        """Function to generate the declariation and initialisation of the
+        temporary variables used for reduction for loop profiling.
+        :param value: String represeting the initial value of temporary
```
Comment (FabioLuporini) [inline devito/profiler.py]: Leave a blank line here before ``:param value: ...``

File: devito/profiler.py [hunk 9]
```diff
@@ -1,30 +1,86 @@
...
+        self.oi = {}
+        self._timings = None
+        self._flops = None
+        self.flops_defaults = {}
 
-    def add_profiling(self, block, name, omp_flag=None):
+    def get_loop_reduction(self, op, variables):
+        """Function to generate reduction pragma, used for profiling under
+        OpenMP settings.
+        :param op: The reduction operator.
```
Comment (FabioLuporini) [inline devito/profiler.py]: Leave a blank line here before ``:param op: ...``

File: devito/profiler.py [hunk 10]
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
...
-    def get_timings_class(self):
+    def get_oi_and_flops(self, name, code, size):
+        """Calculates the total operation intensity of the code provided.
+        If needed, lets the C code calculate it.
+
+        :param name: The name of the field to be populated
+        :param code: The code to be profiled.
+        """
+        if name not in self.oi:
+            self.oi[name] = 0
```
Comment (FabioLuporini) [inline devito/profiler.py]: Consider using a ``defaultdict(int)`` instead of a ``{}`` ?

``from collections import defaultdict``

File: devito/profiler.py [hunk 11]
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
...
+        if name not in self.flops_defaults:
+            self.flops_defaults[name] = 0
```
Comment (FabioLuporini) [inline devito/profiler.py]: same concept as above

File: devito/profiler.py [hunk 12]
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
...
+                assign_flops = self._get_assign_flops(elem, loads)
+                self.oi[name] += assign_flops
+                self.flops_defaults[name] += assign_flops
+            elif isinstance(elem, For):
+                for_flops = self._get_for_flops(name, elem, loads)
+                self.oi[name] += for_flops
+            elif isinstance(elem, Block):
+                block_oi, block_flops = self._get_block_oi_and_flops(name, elem, loads)
+                self.oi[name] += block_oi
+                self.flops_defaults[name] += block_flops
```
Comment (FabioLuporini) [inline devito/profiler.py]: What about AugmentedAssignments, such as Incr, Decr (i.e., +=, -=) ... ?

Maybe we should also have a default ``else`` branch which prints a warning if hit ?

File: devito/profiler.py [hunk 13]
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
...
+                    cur_load = ""
+                    brackets += 1
+                    idx += 1
+                elif char is ' ' and brackets == 0 and len(cur_load) > 0:
+                    loads[cur_load] = True
+                    cur_load = ""
+                    idx += 1
+                else:
+                    cur_load += char
+                    idx += 1
```
Comment (FabioLuporini) [inline devito/profiler.py]: Ah, maybe I see now. This looks very hacky. I think we should make a better use of the cgen hierarchy (by suitably handling Assign, Incr, Decr etc)

File: devito/profiler.py [hunk 14]
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
...
         """Returns the timings as a python dictionary
 
         :returns: A dictionary containing the timings
         """
-        if not self.timings:
+        if not self._timings:
+            return {}
+
+        return dict((field, getattr(self._timings, field))
+                    for field, _ in self._timings._fields_)
```
Comment (FabioLuporini) [inline devito/profiler.py]: I'm not sure I understand this, maybe add an example as a comment (perhaps as method doc ?)

File: devito/profiler.py [hunk 15]
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
...
+        """
+        if not self._flops:
             return {}
 
-        return dict((field, getattr(self.timings, field)) for field, _ in self.timings._fields_)
+        return dict((field,
+                     float(
+                         getattr(self._flops, field) + self.flops_defaults[field]
+                     )/self.timings[field]/10**9)
+                    for field, _ in self._flops._fields_)
```
Comment (FabioLuporini) [inline devito/profiler.py]: Same as before

File: devito/profiler.py [hunk 16]
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
...
+
+    @property
+    def flops(self):
+        if self._flops:
+            return dict((field, int(getattr(self._flops, field))) for field, _ in self._flops._fields_)
+        else:
+            return {}
```
Comment (FabioLuporini) [inline devito/profiler.py]: Looks more complicated than it really needs to be ?

**Comment by pp1336** [comment]
I will handle the formatting issue, @vincepandolfo do you want to handle the others since you probably know those code better

File: devito/profiler.py (hunk 16 above)
Comment (pp1336) [inline devito/profiler.py]: I was just following the previous style, i dont mind changing it.

File: devito/profiler.py (hunk 13 above)
Comment (vincepandolfo) [inline devito/profiler.py]: The only `cgen` objects we use now are `For`, `Block` and `Assign`. That's why it is structured this way.

File: devito/profiler.py (hunk 6 above)
Comment (vincepandolfo) [inline devito/profiler.py]: What would better names be? I agree on having the as read-only attributes.

File: devito/profiler.py (hunk 10 above)
Comment (vincepandolfo) [inline devito/profiler.py]: Why? I'm not familiar with `defaultdict`

File: devito/profiler.py (hunk 14 above)
Comment (vincepandolfo) [inline devito/profiler.py]: What's that you don't get?

Comment (FabioLuporini) [inline devito/profiler.py]: ah, the content of the resulting dict

Comment (vincepandolfo) [inline devito/profiler.py]: `self._timings` is a ctypes struct. `self._timings._fields_` contains a list of tuples `(name, type)`. I get the name from there and the value with `getattr`.

It's simply a conversion from a ctypes struct to a `dict`

File: devito/profiler.py (hunk 10 above)
Comment (FabioLuporini) [inline devito/profiler.py]: If you use a ``defaultdict``, an entry which is not in the dict gets assigned a default value (0 in the case of int). Look at this:

```
In [1]: from collections import defaultdict

In [2]: d = defaultdict(int)

In [3]: d[1]
Out[3]: 0

In [4]: d[2] += 1

In [5]: d
Out[5]: defaultdict(<type 'int'>, {1: 0, 2: 1})
```

Similar behaviour with a normal dict and using the ``setdefault`` method ({}.setdefault.__doc__ is quite self-explanatory)

**Comment by FabioLuporini** [comment]
@pp1336 we need to merge this ASAP. Please join slack.

File: devito/profiler.py (hunk 6 above)
Comment (FabioLuporini) [inline devito/profiler.py]: Maybe ``_C_timings``, ``_C_oi``, ``_C_flops``, ``_C_temporary``. Plus add a comment that says that attributes starting with ``_C_`` are simply used for code generation ?

File: devito/profiler.py (hunk 13 above)
Comment (FabioLuporini) [inline devito/profiler.py]: The method is still hacky, though

The right way of doing this, I think, is to have a visitor that traverses cgen expressions, not analysing strings textually

Comment (vincepandolfo) [inline devito/profiler.py]: The string is the only thing I get from cgen. There's no expression.

**Comment by pp1336** [comment]
vincepandolfo and I have addressed most comments. We agree that we are good with the profiler as it is now, and that the profiler might require further improvements in the future, when we have time.

**Comment by pp1336** [comment]
OK so we are just waiting 81 to be merge, then we will rebase this branch and merge.

**Comment by pp1336** [comment]
Looks like all tests under Linux passed, failed due to flake8. All tests for mac failed.
